### DISCUSSION TITLE
OTBR unable to set Static " OMR Prefix Address "

### DISCUSSION THREAD

**Discussion by Hariprabhu13** [openthread/openthread#9761] [Q&A]
Hi all,
          OT-BR-POSIX is running in Raspberry Pi with the nRF52840 Dongle.

          I have facing the Issue for Setting the On-Mesh Prefix  Address as Statically for Communicating to the Client.

         In OT-CTL it is generating randomly and I changing the IPv6 routable address in the Client once the Device get Reset.
         
         In the OT-BR-POSIX where to set the Default Routable OMR prefix Address to communicate

         So, Suggest me a correct way to set static IPv6 address in the OT-BR-POSIX

Thankyou.

**Comment by Semi-Bin-Blain** [comment]
Maybe the following URL can help:
[prefix command](https://github.com/openthread/openthread/tree/main/src/cli#prefix)
[OpenThread IPv6 Addressing](https://openthread.io/guides/thread-primer/ipv6-addressing)

**Reply by Hariprabhu13** [reply]
Hi,
     I have assigned it in the CLI command using "Prefix add", But the LSB 64 bits are generating Randomly

     I can't fix the SLAAC routable address stable. Once the device get turn off and on it is Changing the Address and  if there is any function available to fix it Static IP.

**Reply by jwhui** [reply]
I would suggest not relying on a static IPv6 address - such a solution will not be resilient to changing network infrastructure configurations.

**Reply by Hariprabhu13** [reply]
Hi @jwhui,
                  Setting the static IPv6 is not possible means,
                  
                  Im using the raspberrypi as Border Router and ESP32C6 as the End Device in that the IP address generation, I need only the SLAAC address for communicating the SRP server and client from pinging the IPV6 outside the Device in Same Network
                  
                  ipaddr -v

                  fdc8:c268:9c4:1:42d8:bfdb:a22a:737b origin:slaac
                  fd99:61fa:3b76:3296:0:ff:fe00:dc01 origin:thread
                  fd99:61fa:3b76:3296:ffd5:ece0:44e1:e83c origin:thread
                  fe80:0:0:0:a4ce:6c7:a3e6:3150 origin:thread

                  For the Every Address get from the Thread Network having the Function without using CLI, But for getting SLAAC address there is No CLI and Function to get the Address 
                  
                  So, please suggest me to get the SLAAC address Function in openthread API.

**Reply by jwhui** [reply]
DNS-SD is typically used for discovering IPv6 addresses.

**Comment by Hariprabhu13** [comment]
Hi @jwhui 
                  Thanks for your reply, 
                  I have a doubt in list of ip address shown in CLI

                  In CLI, "ipaddr" command list out the IPv6 Address of the Device
                  Log:
                  I (528) OT_STATE: netif up
I(548) OPENTHREAD:[N] Mle-----------: Role detached -> child
I (628) OT_STATE: Set dns server address: FDC8:C268:9C4:2::808:808
> state

child
Done
> ipaddr -v

fdc8:c268:9c4:1:998b:350d:66c6:3969 origin:slaac
fde3:14b5:cc2a:3aa8:0:ff:fe00:dc02 origin:thread
fde3:14b5:cc2a:3aa8:ce58:59a7:301e:5f63 origin:thread
fe80:0:0:0:b413:6938:295a:c12c origin:thread
Done
>

           But, I have written a API function to get all the List of IPv6 Address in the CMD line it shows
            Log : 
            I (468) AU79_CLI: The Network Key is Initialised Successfully

I(478) OPENTHREAD:[N] Mle-----------: Role disabled -> detached
> I (498) OPENTHREAD: OpenThread attached to netif
I (498) AU79_CLI: The IPV6 Address : fde3:14b5:cc2a:3aa8:0:ff:fe00:dc02
I (498) AU79_CLI: The IPV6 Address : fde3:14b5:cc2a:3aa8:ce58:59a7:301e:5f63
I (508) AU79_CLI: The IPV6 Address : fe80:0:0:0:b413:6938:295a:c12c
I (518) AU79_CLI: Device State : detached

netif1: ot

netif1: lo
lo inet6: ::1 16

I (528) OT_STATE: netif up
I(548) OPENTHREAD:[N] Mle-----------: Role detached -> child
I (628) OT_STATE: Set dns server address: FDC8:C268:9C4:2::808:808
> state

           So , clarify me that API only listing the Thread origin Address, It does not getting assigned SLAAC address to the API function

           But using CLI, it prints all the address , but I can't get it throw the API function
            
           // Get Thread Network IPv6 Address
        const otNetifAddress *ipaddr = otIp6GetUnicastAddresses(cInstance);

        char ip[OT_IP6_ADDRESS_STRING_SIZE];

        for(const otNetifAddress *ucast = ipaddr; ucast; ucast = ucast->mNext)
        {
                // Print IPv6 address Assigned to Thread Device
                otIp6AddressToString((const otIp6Address *)ucast, ip, sizeof(ip));

                ESP_LOGI(TAG1, "The IPV6 Address : %s", ip);
        }

        So, please clarify how to get the assigned SLAAC address in the API Function and the Openthread CLI is Initiated after that how to lock the CLI API's to execute the other API Functions in it , please provide any API is available for stop CLI process and Execute my API Functions

**Reply by abtink** [reply]
The [`otIp6GetUnicastAddresses()`](https://openthread.io/reference/group/api-ip6#otip6getunicastaddresses) does return all unicast addresses independent of their origin, so SLAAC addresses are included in the list.

I think you may be calling the API too early when device is still "detached".

```
I (498) OPENTHREAD: OpenThread attached to netif
I (498) AU79_CLI: The IPV6 Address : fde3:14b5:cc2a:3aa8:0:ff:fe00:dc02
I (498) AU79_CLI: The IPV6 Address : fde3:14b5:cc2a:3aa8:ce58:59a7:301e:5f63
I (508) AU79_CLI: The IPV6 Address : fe80:0:0:0:b413:6938:295a:c12c
I (518) AU79_CLI: Device State : detached
```

**Comment by Hariprabhu13** [comment]
Hi @abtink 
                   Thanks, for reply now the State is turned to Child, I'm Getting the SLAAC Address...

**Comment by Hariprabhu13** [comment]
Hi @abtink ,
                    I have doubt on openthread border router in ot-br-posix to attaching a Device's , the Child and Neighbor table shows the list and Address of Devices. But , When I disconnected the Device's It takes a time to remove the list of attached device list.

                    The " history neighbor and neighbor lists " in OT-CLI also show's the Added device details immediately, but while it is taking time to show the removed device detail at the time.
                    
                    What is the Solution to know the device details in OT-BR-POSIX to know the device is added and removed.

**Reply by jwhui** [reply]
State removal in network protocols is typically done using timeouts. As a result, the timeout must expire before a neighbor is detected as having left of disappeared.

BTW, this question seems unrelated to the original question above. Please create a new discussion for new questions or topipcs.
